### ISSUE TITLE
Red Junglefowl training data seem to have included some "boars"? -

### ISSUE THREAD

**Issue by HugoMarkoff** [google/cameratrapai#44]
So I used my https://github.com/AI-EcoNet/Cameratrap_Image_Downloader to download 3000 images of Red Junglefowl. Especially on the Red Junglefowl data, I noticed I downloaded quite a few images of other animals (with no Red Junglefowl). So I can only assume these images has been labeled incorrectly, and most likely used for training Species Net. 

Basically - I get a lot of boars classified as Red Junglefowl, from this specific data. 

If you have more questions - please reach out 

![Image](https://github.com/user-attachments/assets/1ffbf207-5f02-4c9f-87b9-27846a6f2c4d)

<img width="1698" height="857" alt="Image" src="https://github.com/user-attachments/assets/908f8d71-7f84-4b93-86cc-63383fd872d1" />

![Image](https://github.com/user-attachments/assets/288ad0ba-e7b0-423a-9979-620a3d60d5e7)

**Comment by stefanistrate** [comment]
I can confirm that some of those incorrectly labelled images made it into the training set for SpeciesNet. Thanks for flagging this out and let out know if you discover more.

**Comment by HugoMarkoff** [comment]
I will let you know if I find any similar cases. So far this has been the only class really standing out running Species Net. However, if much of the data has been used from Lila with the labels from the COCO.json files from there, I have spotted a few other classes which seem to "overlap" with wrong data (Actually quite a lot of cases).

Lets "just" take wolf as example. 

I used the CSV from https://lila.science/taxonomy-mapping-for-camera-trap-data-sets/ (and some AI magic) to match all species across all the datasets. While having both "Domestic Dog" and "Gray wolf" as two separate classes in Species Net, I noticed quite some Domestic dogs sneak into the Gray Wolf folder. (WCS Camera Traps,canis lupus,species,canis lupus,gray wolf,"[(42048, 'species', 'canis lupus', ['gray wolf'])) <- this should be Domestic Dog so if you used the same mapping be aware of this. 

And at the same time.. you will have images labeled with the class: "category_id": 73 (canis lupus) looking like this:

![Image](https://github.com/user-attachments/assets/e1851b07-5ad6-485d-95c7-698bb8121471)

![Image](https://github.com/user-attachments/assets/bdf98228-ae42-42c5-afb4-6c018c121ed2)

![Image](https://github.com/user-attachments/assets/c45e282b-fc56-46ae-9c32-5b97a563b544)

![Image](https://github.com/user-attachments/assets/c6074c22-9945-4da6-a579-13aafec3789c)

While most others do actually have dogs in them 

Also a few Bob cats labeled as wolves in idaho-camera-traps (I backtracked 3 of the images to manually verify, and they are in the COCO.json of idaho-camera-traps labeled with "category_id": 38, which is gray wolf)

<img width="1658" height="659" alt="Image" src="https://github.com/user-attachments/assets/dafc94e5-65b9-4b45-9ac4-438b16e54af2" />

I see this quite a lot, while some datasets are clearly better labeled and sorted than others, there are MANY cases across the data which are incorrectly labeled. 

Maybe consider sanitizing the data before next re-training, taking 1 class at the time, actually run it through MD + SN and "verify" some of the labels. I mean you will probably use Wildlife Insights for this tasks, but if you need a convenient tool which also groups up based on similarities to spot these "groups" of "incorrect" animals you could always use Animal Detect (wink wink)

<img width="1663" height="641" alt="Image" src="https://github.com/user-attachments/assets/7b93c049-220e-463e-8a19-15de7ca6846a" />

**Comment by agentmorris** [comment]
Thanks for all this debugging, @HugoMarkoff !  The red junglefowl issue looks systematic within SWG Camera Traps, so I will manually review those and correct them where I can, and where I can't identify the species, I'll call it "animal".  I don't know when I will get to this, or when we will re-train SpeciesNet.

In general, I have not historically tried to chase down every incorrect label on LILA: it looks like a lot when we call them out here, but these are very large datasets, and there is no such thing as perfect data, so we expect a certain amount of mislabeling.  Deep learning is surprisingly robust!  But I would like to fix the systematic ones... e.g. if there are 10 bobcats labeled as wolves, I'll probably leave it, if there are 1000 bobcats labeled as wolves, I'd like to fix it.

@HugoMarkoff, if you have a list of images paths (from the .csv file) that you believe are incorrect, please send me the URLs, along with the incorrect/probably-correct labels.  I will figure out an efficient UI for reviewing those.  I will probably only make corrections if there are at least 100 of the same error within a single dataset.

I think it would be a *fantastic* project for someone in the community to run SpeciesNet on all of LILA, and look for high-confidence disagreements with the ground truth that are taxonomically nonsensical.  I think it will be hard to catch dog/wolf confusion this way (because if you flag every time SpeciesNet predicts the wrong species in the same family, you'll be reviewing hundreds of thousands of images), but one might catch a good number of pig/junglefowl errors this way (despite the fact that SpeciesNet was trained on this data).  Ditto for blank/animal confusion, particularly in WCS Camera Traps.  I have already run MegaDetector on all the camera trap data on LILA ([results](https://lila.science/megadetector-results-for-camera-trap-datasets/)), including [repeat detection elimination](https://github.com/agentmorris/MegaDetector/tree/main/megadetector/postprocessing/repeat_detection_elimination), so it shouldn't be necessary to run MD everywhere, just SpeciesNet.

We'll leave this issue open until I fix the junglefowl labels (or decide not to).

**Comment by HugoMarkoff** [comment]
I will let you know if I find datasets with significant errors (so excluding cases like 10-20 bobcats sneaking into the wolf data of 2600+ images) 

Deep learning is indeed doing a good job with just a few outliers of other classes. I will probably have a small break of using LILA data for the next coming month(s) as I got most of the data I need now. But will for sure share if I find something else significant 😊 

A general observation I have seen quite a lot throughout much of the images I have processed is the occurrence of “domestic cattle” from completely rubbish images and completely different classes, like crows in the distance. They seem to sneak up quite often compared to other false negative species-specific classes. It is however quite random and not systematic – just an observation

**Comment by agentmorris** [comment]
Thank you again for finding this issue, @HugoMarkoff !  I'm finally closing this, because although it's not fixed in SpeciesNet per se, it's fixed in the source and will either be ingested into SpeciesNet in the next training round or the round after that.

For posterity, here is what happened here, starting with some quick background for anyone reading this issue who isn't familiar with camera trap data:

* Camera traps are typically configured to capture 3-10 images when the motion sensor is triggered.  That group of images (from a single trigger) may be called a "sequence", "event", "episode", or "burst".  We'll use the term "sequence".
* It's almost never ecologically relevant which images in a sequence have which species in them, or which are empty.  So in most camera trap datasets, the ecologists doing the labeling give the same label to all the images in a sequence.  E.g. if there is a lion in the first two frames of a three-frame sequence, but the lion is out of view in the third frame, it's common for the three images to just be labeled as lion/lion/lion.  I'll call this "sequence-level labeling".
* The opposite would be "image-level labeling", where every image is labeled accurately.  This has no ecological value over sequence-level labeling, and it's a huge pain, so it's rare.  [SWG Camera Traps](https://lila.science/datasets/swg-camera-traps/) - the dataset from which this pig/junglefowl confusion stemmed - is labeled at the sequence level.
* The [data format we use for LILA datasets](http://lila.science/coco-camera-traps) has an optional flag called "sequence_level_annotation", which indicates whether an annotation applies to an image or a sequence.  SWG Camera Traps (correctly) sets this flag to "True" for every annotation.  So far so good.
* 99.9999% of the time when "sequence_level_annotation" is set to "True", the same labels are provided for every image in the sequence.  E.g., if there's a lion in frame 1 of a three-frame sequence, and a zebra in frame 3, and the middle frame is empty, and sequence_level_annotation is True, all three would typically be called "zebra,lion".  But this isn't actually enforced or required in the format spec.
* SWG Camera Traps had a few instances where sequence_level_annotation is (correctly) set to True, and sequences contained both red junglefowl and Eurasian wild pigs, but for whatever reason, each label was applied to just a subset of images in the sequence (and not necessarily the images within the sequence where those species occur).  Some of these were very long sequences (hundreds of images).  There is an 84% chance this is my fault (I built the original .json file for this dataset a zillion years ago), but we'll never really know how this happened, i.e., why those labels weren't propagated to every image in the sequence.
* This isn't exactly <i>wrong</i>, in theory someone using this data (in this case, us (Google), when getting ready to train SpeciesNet), might propagate all the labels in a sequence to every image in the sequence if sequence_level_annotation is True.  But that's not very intuitive, so this is more or less a dataset bug, and in any case, we didn't do this.

So, I just propagated all labels in each sequence to all images in the sequence, re-generated the .json file for this dataset, and re-generated the [big .csv file](https://lila.science/taxonomy-mapping-for-camera-trap-data-sets/) that represents all LILA camera trap data.  The next time we train SpeciesNet (or possibly the time after that, depending on timing), we'll see every image in these sequences as a multi-species image (where it's not clear which boxes are which species), and we'll throw them out.  Voila, no more pigs called junglefowl.

I checked every other dataset on LILA for cases where this particular pathology occurs, and it doesn't occur anywhere else.  That doesn't mean every other dataset is perfect; as per above, all datasets have bugs.  But this particular issue re: sequence_level_annotation and non-propagated labels doesn't occur in other datasets.

Thanks again for reporting this!
